CI: do not cancel concurrent builds on the same branch

Kobzol opened the pull request:
Do not cancel concurrent builds on the same branch (outside of PRs).

Instead, only cancel them if the builds have the same commit SHA.

From the [documentation](https://docs.github.com/en/actions/learn-github-actions/contexts#github-context):
> The commit SHA that triggered the workflow. The value of this commit SHA depends on the event that triggered the workflow. For more information, see "[Events that trigger workflows](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows)." For example, ffac537e6cbbf934b08745a378932722df287a53.

Fixes: https://github.com/rust-lang/rust/pull/112955#discussion_r1242273658

r? @pietroalbini

Kobzol committed:
CI: do not cancel concurrent builds on the same branch

Instead, only cancel them if the builds have the same commit SHA.

Kobzol commented:
To stay on the safe side, I will experiment on another repo if this actually does what we want.

pietroalbini commented on src/ci/github-actions/ci.yml:
```diff
@@ -301,7 +301,7 @@ defaults:
   # For a given workflow, if we push to the same PR, cancel all previous builds on that PR.
   # If the push is not attached to a PR, we will cancel all builds related to the same commit SHA.
-  group: ${{ github.workflow }}-${{ github.event.pull_request.number || github.ref }}
+  group: ${{ github.workflow }}-${{ github.event.pull_request.number || github.sha }}
```
I think this is also not the behavior we want, because we *do* want to cancel concurrent auto builds, while we don't want to cancel concurrent try builds. I think we can use [this to check which branch we're in](https://docs.github.com/en/actions/learn-github-actions/expressions#example).

Kobzol replied:
Oh, I thought that auto builds happen in the same way as try builds. But yeah, it makes sense to "serialize" the auto branch. I guess that bors makes sure that it doesn't run in parallel at the moment?

Kobzol replied:
I have done a few more experiments and I suggest this expression as the concurrency key:
```yaml
${{ (github.ref == 'refs/heads/try' && github.sha) || github.ref }}
```
By default, we only allow one CI per branch. This will work both for auto and for PRs, because for PRs the ref will look like this: `refs/pull/<pr_number>/merge`. On top of that, we add an explicit exception for the `try` branch, which will use the commit SHA as the concurrency key, and thus it will allow concurrent builds.

pietroalbini replied:
Sounds good.

Kobzol committed:
CI: do not cancel concurrent builds on the same branch

Add an exception for try and try-perf branches to enable concurrent try builds and unrolled rollup builds.

Kobzol commented:
I implemented the idea from my [comment](https://github.com/rust-lang/rust/pull/113059#discussion_r1245292438), and added exception for `try` and `try-perf` branches.

pietroalbini commented:
@bors r+ p=1000

bors commented:
:pushpin: Commit 32428ab5f37b56314715b23b14dd700510d84f1e has been approved by `pietroalbini`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=pietroalbini 32428ab5f37b56314715b23b14dd700510d84f1e -->
<!-- homu: {"type":"Approved","sha":"32428ab5f37b56314715b23b14dd700510d84f1e","approver":"pietroalbini","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 32428ab5f37b56314715b23b14dd700510d84f1e with merge b23a5add09283360bb3b8eaa4292ce3366e62615...
<!-- homu: {"type":"BuildStarted","head_sha":"32428ab5f37b56314715b23b14dd700510d84f1e","merge_sha":"b23a5add09283360bb3b8eaa4292ce3366e62615"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/5415390336/jobs/9846658003)
Approved by: pietroalbini
Pushing b23a5add09283360bb3b8eaa4292ce3366e62615 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"pietroalbini","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/5415390336/jobs/9846658003"},"merge_sha":"b23a5add09283360bb3b8eaa4292ce3366e62615"} -->

klensy commented on src/ci/github-actions/ci.yml:
```diff
@@ -299,9 +299,10 @@ defaults:
-  # For a given workflow, if we push to the same PR, cancel all previous builds on that PR.
-  # If the push is not attached to a PR, we will cancel all builds related to the same commit SHA.
-  group: ${{ github.workflow }}-${{ github.event.pull_request.number || github.ref }}
+  # For a given workflow, if we push to the same branch, cancel all previous builds on that branch.
+  # We add an exception for try builds (try branch) and unrolled rollup builds (try-perf), which
+  # are all triggered on the same branch, but which should be able to run concurrently.
+  group: ${{ ((github.ref == 'refs/heads/try' || github.ref == 'refs/heads/try-perf') && github.sha) || github.ref }}
```
Is removing `${{ github.workflow }}` part of key is expected? As now this is not equal comment `For a given workflow, if ...`

Kobzol replied:
We have only one workflow that is relevant here currently, but you're right that the comment is outdated. I suppose that for future-proofing it wouldn't be a bad idea to include the workflow in the group.

rust-timer commented:
Finished benchmarking commit (b23a5add09283360bb3b8eaa4292ce3366e62615): [comparison URL](https://perf.rust-lang.org/compare.html?start=a20a04e5d68e85935539c16c945f1947559ad3ce&end=b23a5add09283360bb3b8eaa4292ce3366e62615&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=a20a04e5d68e85935539c16c945f1947559ad3ce&end=b23a5add09283360bb3b8eaa4292ce3366e62615&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=a20a04e5d68e85935539c16c945f1947559ad3ce&end=b23a5add09283360bb3b8eaa4292ce3366e62615&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.5%  | [1.5%, 3.6%]   | 2     |
| Regressions ❌ <br /> (secondary)  | 0.8%  | [0.8%, 0.8%]   | 1     |
| Improvements ✅ <br /> (primary)   | -0.6% | [-0.8%, -0.5%] | 2     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | 0.9%  | [-0.8%, 3.6%]  | 4     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=a20a04e5d68e85935539c16c945f1947559ad3ce&end=b23a5add09283360bb3b8eaa4292ce3366e62615&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.4% | [0.4%, 0.4%] | 1     |
| Regressions ❌ <br /> (secondary)  | -    | -            | 0     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.4% | [0.4%, 0.4%] | 1     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=a20a04e5d68e85935539c16c945f1947559ad3ce&end=b23a5add09283360bb3b8eaa4292ce3366e62615&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 660.963s -> 660.993s (0.00%)
